lib/mongo/server/connection_pool/available_stack.rb

```diff
@@ -300,31 +288,83 @@ def close_stale_sockets!
           end
 
           mutex.synchronize do
-            num_checked_out = pool_size - queue_size
+            num_checked_out = pool_size - stack_size
             min_size_delta = [(min_size - num_checked_out), 0].max
 
             to_refresh.each do |connection|
-              if queue.include?(connection)
+              if connections.include?(connection)
                 connection.disconnect!
-                if queue.index(connection) < min_size_delta
+                if connections.index(connection) < min_size_delta
                   begin; connection.connect!; rescue; end
                 end
               end
             end
           end
         ensure
-          check_count_invariants
+          check_count_invariants(false)
+        end
+
+        def connection_removed
+          @pool_size -= 1
+          if @pool_size < 0
+            # This should never happen
+            log_warn("ConnectionPool::AvailableStack: unexpected push")
```

Comment:
This message needs to be updated because this method can be called as part of a sequence of operations not involving a push.

Reply:
I've updated it; let me know if you prefer a different error message than the one I picked